src/lib.rs

```diff
@@ -1861,22 +1867,17 @@ impl Connection {
 
                 let frame = frame::Frame::MaxStreamData {
                     stream_id,
-                    max: stream.recv.update_max_data() as u64,
+                    max: stream.recv.max_data_next(),
                 };
 
-                self.streams.mark_almost_full(stream_id, false);
+                if push_frame_to_pkt!(frames, frame, payload_len, left) {
+                    stream.recv.update_max_data();
 
-                if frame.wire_len() > left {
-                    break;
+                    ack_eliciting = true;
+                    in_flight = true;
                 }
 
-                payload_len += frame.wire_len();
-                left -= frame.wire_len();
-
-                frames.push(frame);
-
-                ack_eliciting = true;
-                in_flight = true;
+                self.streams.mark_almost_full(stream_id, false);
```

Comment:
This should be inside the `if`, otherwise we won't retry sending the frame if it failed this time.